### ISSUE TITLE
Negative Loss Computations

### ISSUE THREAD

**Issue by MrZilinXiao** [haon-chen/MoCa#2]
Hi, Thanks for the great work!

I am looking at the https://github.com/haon-chen/MoCa/blob/fb3ac280cb69ddb0e769e3903a77e4e6441aa278/src/model.py#L276, where the target is set to diagonal but scores are organized in [B, 1 + neg_ratio]. So from my understanding, the target should be set to all zeros. Am I missing something?

**Comment by haon-chen** [comment]
Hi @MrZilinXiao, thanks for the careful reading!

`compute_similarity(q, t)` returns the *full* pair-wise dot-product matrix `𝑆∈ℝ^{B×B}` (equivalent to `torch.matmul(q, t.T)`).

```python
pos_scores = self.compute_similarity(all_qry_reps, all_tgt_reps)  # (B, B)
pos_scores = pos_scores.view(B, -1)                               # still (B, B)
```

So after concatenating the query-specific negatives

```python
neg_scores  # (B, neg_ratio)
scores = torch.cat([pos_scores, neg_scores], dim=1)  # (B, B + neg_ratio)
```

each row now contains

* **B in-batch positives** – the `i`-th column is the true match for query `i`;
* `neg_ratio` hard negatives drawn with `neg_reps`.

Accordingly, the ground-truth label for query `i` is simply the diagonal index `i`, produced by

```python
target = torch.arange(B, device=scores.device)  # 0,1,…,B-1
```

(The extra `* (all_qry_reps.size(0) // all_tgt_reps.size(0))` keeps indices consistent when we gather tensors across GPUs; on a single GPU this factor is 1.)

If we had used only one positive per row (`[B, 1 + neg_ratio]`) then the target would indeed be all zeros, but we purposely keep all in-batch pairs to leverage the CLIP-style InfoNCE loss, which empirically gives better performance.

Hope this clarifies the design—let me know if anything is still unclear!

**Comment by MrZilinXiao** [comment]
Thanks for the quick and clear response, which clarifies nearly all of my questions!

One last question, I thought `all_qry_reps` and `all_tgt_reps` should be of the same size as they all went through all_gather -- what will cause them to have different B?

**Comment by haon-chen** [comment]
Hi @MrZilinXiao,
You are right, they should have the same B.

**Comment by MrZilinXiao** [comment]
> Hi [@MrZilinXiao](https://github.com/MrZilinXiao), You are right, they should have the same B.

Right. Then `(all_qry_reps.size(0) // all_tgt_reps.size(0)` will always equal to 1 -- which seems to make no sense to make an additional efforts on that line.

**Comment by haon-chen** [comment]
Hi @MrZilinXiao,
Thanks for the careful reading!
This line was meant to handle different batch structures, but it serves no purpose in the current implementation.
